File: src/GnOuGo.Flow.Core/Runtime/Executors/LlmCallExecutor.cs
using System.Text.Json.Nodes;
using Microsoft.Extensions.Logging;
using GnOuGo.AI.Core;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Calls an LLM via the configured ILLMClient.
/// </summary>
public sealed class LlmCallExecutor : IStepExecutor
{
    public string StepType => "llm.call";

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The input object is malformed or the required `prompt` field is missing, and no runtime default `model` is available."),
        new(ErrorCodes.LlmTimeout, true, "The LLM request timed out. This is retryable and is a good candidate for `retry`."),
        new(ErrorCodes.LlmNetwork, true, "A transient network failure occurred while calling the LLM provider."),
        new(ErrorCodes.LlmNetwork, false, "The LLM client is not configured or the provider failed in a non-transient way."),
        new(ErrorCodes.LlmSchema, false, "structured_output was requested but the LLM returned a response that could not be parsed as valid JSON.")
    };

    public string DslSnippet => """
        ### llm.call — Call a language model
        IMPORTANT: use `prompt` (NOT `messages`). `prompt` is REQUIRED. `model` is required unless the runtime injects a default model.
        IMPORTANT: `temperature` and `reasoning` are optional overrides. Omit them unless the task explicitly needs them; the runtime applies defaults and removes unsupported parameters based on model capabilities.
        Basic call:
        ```yaml
        - id: summarize
          type: llm.call
          input:
            model: gpt-4                        # optional when runtime defaults are configured
            prompt: "Summarize: ${data.steps.prev.text}"  # required — plain string
            system: "You are a helpful assistant."  # optional
            temperature: 0.7                     # optional override; omit by default
            reasoning: high                      # optional override; omit by default
            max_tokens: 2048                     # optional
        ```
        Structured output:
        IMPORTANT for `strict: true` (OpenAI/GitHub Models response_format json_schema):
        - Every schema object with `properties` MUST have `required` listing EVERY key from `properties`.
        - Do NOT list only the fields that feel mandatory; strict mode rejects omitted property names.
        - Optional fields must still be listed in `required`; represent them as nullable with `anyOf: [{ type: <type> }, { type: "null" }]`.
        - Add `additionalProperties: false` on every object schema for portability. The OpenAI provider also patches it automatically in strict mode.
        ```yaml
        - id: classify
          type: llm.call
          input:
            model: gpt-4
            prompt: "Classify this ticket and return JSON"
            structured_output:
              schema_inline:
                type: object
                properties:
                  category: { type: string }
                  priority: { type: string }
                  notes:
                    anyOf:
                      - type: string
                      - type: "null"
                required: [category, priority, notes]   # every property above is listed, including nullable optional fields
                additionalProperties: false
              strict: true                       # optional
        ```
        You can also use `structured_output.schema_ref` instead of `schema_inline`.
        Output: `{ text: "...", json?: {...}, usage: { prompt_tokens, completion_tokens, total_tokens }, meta: { model } }`
        """;

    public async Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var llmClient = ctx.Engine.LLMClient
            ?? throw new WorkflowRuntimeException(ErrorCodes.LlmNetwork, "No LLM client configured");

        var input = ctx.Engine.GetResolvedInput(ctx) as JsonObject
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "llm.call input must be object");

        var requestedProvider = input["provider"]?.GetValue<string>();
        var requestedModel = input["model"]?.GetValue<string>();
        var (provider, model) = ctx.Engine.ResolveLlmTarget(requestedProvider, requestedModel);
        if (string.IsNullOrWhiteSpace(model))
        {
            throw new WorkflowRuntimeException(
                ErrorCodes.InputValidation,
                "llm.call requires 'model' unless WorkflowEngine.LlmDefaults.Model is configured");
        }

        var prompt = input["prompt"]?.GetValue<string>()
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "llm.call requires 'prompt'");

        var request = new LLMRequest
        {
            Provider = provider,
            Model = model,
            Prompt = prompt,
        };

        // ── Telemetry: record request attributes ──
        ctx.SetTelemetryAttribute("gen_ai.operation.name", "chat");
        ctx.SetTelemetryAttribute("gen_ai.system", request.Provider ?? "default");
        ctx.SetTelemetryAttribute("gen_ai.request.model", model);

        if (ctx.Limits.LogStepContent)
        {
            ctx.AddTelemetryEvent("gen_ai.content.prompt", new[]
            {
                new KeyValuePair<string, object?>("gen_ai.prompt", prompt),
                new KeyValuePair<string, object?>("prompt.role", "user")
            });
        }

        if (input.TryGetPropertyValue("temperature", out var temp) && temp != null)
        {
            request.Temperature = ExpressionEvaluator.GetNumber(temp);
            ctx.SetTelemetryAttribute("gen_ai.request.temperature", request.Temperature);
        }

        // Reasoning / thinking effort: "minimal"|"low"|"medium"|"high"|"max"|"auto".
        // When omitted, providers fall back to their own defaults ("auto").
        if (input.TryGetPropertyValue("reasoning", out var reasoningNode) && reasoningNode != null)
        {
            var reasoning = reasoningNode.GetValue<string>();
            if (!string.IsNullOrWhiteSpace(reasoning))
            {
                request.Reasoning = reasoning;
                ctx.SetTelemetryAttribute("gen_ai.request.reasoning_effort", reasoning);
            }
        }

        // Structured output
        var structuredOutput = input["structured_output"] as JsonObject;
        if (structuredOutput != null)
        {
            request.StructuredOutputSchema = structuredOutput["schema_inline"] ?? structuredOutput["schema_ref"];
            if (structuredOutput.TryGetPropertyValue("strict", out var s) && s != null)
                request.StructuredOutputStrict = s.GetValue<bool>();
        }

        // Max output tokens
        if (input.TryGetPropertyValue("max_tokens", out var maxTokensNode) && maxTokensNode != null)
        {
            request.MaxTokens = (int)ExpressionEvaluator.GetNumber(maxTokensNode);
            ctx.SetTelemetryAttribute("gen_ai.request.max_tokens", request.MaxTokens);
        }

        // ── Thinking: signal that we are calling the LLM ──
        ctx.AddTelemetryEvent("gnougo-flow.step.thinking", new[]
        {
            new KeyValuePair<string, object?>("gnougo-flow.thinking.message", $"Calling LLM ({model})…"),
            new KeyValuePair<string, object?>("gnougo-flow.thinking.level", "thinking")
        });

        try
        {
            var response = await llmClient.CallAsync(request, ct);


            // ── Thinking: preview first 120 chars of the LLM response ──
            if (!string.IsNullOrWhiteSpace(response.Text))
            {
                var preview = response.Text.Length > 120
                    ? response.Text[..120] + "…"
                    : response.Text;
                ctx.AddTelemetryEvent("gnougo-flow.step.thinking", new[]
                {
                    new KeyValuePair<string, object?>("gnougo-flow.thinking.message", preview),
                    new KeyValuePair<string, object?>("gnougo-flow.thinking.level", "response")
                });
            }

            // ── Telemetry: record response attributes ──
            var finishReason = response.ToolCalls?.Count > 0 ? "tool_calls" : "stop";
            ctx.SetTelemetryAttribute("gen_ai.response.model", model);
            ctx.SetTelemetryAttribute("gen_ai.response.finish_reason", finishReason);
            long? inputTokens = null;
            long? outputTokens = null;
            if (response.Usage is JsonObject usage)
            {
                if (usage.TryGetPropertyValue("prompt_tokens", out var pt) && pt != null)
                    inputTokens = pt.GetValue<int>();
                else if (usage.TryGetPropertyValue("input_tokens", out var it) && it != null)
                    inputTokens = it.GetValue<int>();

                if (usage.TryGetPropertyValue("completion_tokens", out var ct2) && ct2 != null)
                    outputTokens = ct2.GetValue<int>();
                else if (usage.TryGetPropertyValue("output_tokens", out var ot) && ot != null)
                    outputTokens = ot.GetValue<int>();

                if (inputTokens.HasValue)
                    ctx.SetTelemetryAttribute("gen_ai.usage.input_tokens", inputTokens.Value);
                if (outputTokens.HasValue)
                    ctx.SetTelemetryAttribute("gen_ai.usage.output_tokens", outputTokens.Value);

                if (usage.TryGetPropertyValue("total_tokens", out var tt) && tt != null)
                    ctx.SetTelemetryAttribute("gen_ai.usage.total_tokens", tt.GetValue<int>());

                var estimatedCost = ModelMetadataCatalog.EstimateCost(model, inputTokens, outputTokens, providerType: provider);
                if (estimatedCost.HasValue)
                    ctx.SetTelemetryAttribute("gen_ai.usage.cost", (double)estimatedCost.Value);
            }

            if (ctx.Limits.LogStepContent && (!string.IsNullOrWhiteSpace(response.Text) || response.Json != null))
            {
                ctx.AddTelemetryEvent("gen_ai.content.completion", new[]
                {
                    new KeyValuePair<string, object?>("gen_ai.completion", !string.IsNullOrWhiteSpace(response.Text)
                        ? response.Text
                        : response.Json?.ToJsonString()),
                    new KeyValuePair<string, object?>("completion.role", "assistant"),
                    new KeyValuePair<string, object?>("completion.finish_reason", finishReason)
                });
            }

            var result = new JsonObject
            {
                ["text"] = response.Text,
                ["meta"] = new JsonObject { ["model"] = model }
            };

            // Structured output: fallback parse response.Text as JSON (parity with Python)
            JsonNode? structuredJson = response.Json;
            if (structuredOutput != null && structuredJson == null && !string.IsNullOrWhiteSpace(response.Text))
            {
                var textToParse = StripMarkdownCodeFences(response.Text);
                try { structuredJson = JsonNode.Parse(textToParse); }
                catch { structuredJson = null; }
            }
            if (structuredOutput != null && structuredJson == null)
                throw new WorkflowRuntimeException(ErrorCodes.LlmSchema,
                    "llm.call structured_output expected valid JSON but the LLM returned an incompatible response", retryable: true);

            if (structuredJson != null)
                result["json"] = structuredJson.DeepClone();
            else if (response.Json != null)
                result["json"] = response.Json.DeepClone();
            if (response.Usage != null)
                result["usage"] = response.Usage.DeepClone();
            if (response.Raw != null)
                result["raw"] = response.Raw.DeepClone();

            return result;
        }
        catch (TimeoutException ex)
        {
            ctx.Engine.Logger.LogError(ex, "llm.call timed out for model '{Model}'", model);
            throw new WorkflowRuntimeException(ErrorCodes.LlmTimeout, ex.Message, retryable: true);
        }
        catch (HttpRequestException ex)
        {
            ctx.Engine.Logger.LogError(ex, "llm.call network error for model '{Model}'", model);
            throw new WorkflowRuntimeException(ErrorCodes.LlmNetwork, ex.Message, retryable: true);
        }
        catch (WorkflowRuntimeException)
        {
            throw;
        }
        catch (Exception ex)
        {
            ctx.Engine.Logger.LogError(ex, "llm.call failed for model '{Model}'", model);
            throw new WorkflowRuntimeException(ErrorCodes.LlmNetwork, $"LLM call failed: {ex.Message}", retryable: false);
        }
    }

    /// <summary>
    /// Strips markdown code fences (```json ... ``` or ``` ... ```) that LLMs
    /// sometimes wrap around structured JSON responses.
    /// </summary>
    internal static string StripMarkdownCodeFences(string text)
    {
        var trimmed = text.Trim();
        if (!trimmed.StartsWith("```"))
            return trimmed;

        // Remove opening fence (```json, ```JSON, ``` etc.)
        var firstNewline = trimmed.IndexOf('\n');
        if (firstNewline < 0)
            return trimmed;

        var body = trimmed[(firstNewline + 1)..];

        // Remove closing fence
        var lastFence = body.LastIndexOf("```", StringComparison.Ordinal);
        if (lastFence >= 0)
            body = body[..lastFence];

        return body.Trim();
    }
}

File: librairies/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/llm_call_executor.py
from __future__ import annotations

import gnougo_flow_core.runtime as _runtime
from gnougo_flow_core.runtime import *  # noqa: F401,F403


def _matches_json_schema(value: Any, schema: Any) -> bool:
    if not isinstance(schema, dict):
        return True

    schema_type = schema.get("type")
    if isinstance(schema_type, list):
        return any(_matches_json_schema(value, {**schema, "type": item}) for item in schema_type)

    if schema_type == "object":
        if not isinstance(value, dict):
            return False
        required = schema.get("required") if isinstance(schema.get("required"), list) else []
        if any(key not in value for key in required):
            return False
        properties = schema.get("properties") if isinstance(schema.get("properties"), dict) else {}
        return all(key not in value or _matches_json_schema(value[key], prop_schema) for key, prop_schema in properties.items())
    if schema_type == "array":
        if not isinstance(value, list):
            return False
        item_schema = schema.get("items")
        return all(_matches_json_schema(item, item_schema) for item in value) if item_schema else True
    if schema_type == "string":
        return isinstance(value, str)
    if schema_type in {"number", "integer"}:
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if schema_type == "boolean":
        return isinstance(value, bool)
    if schema_type == "null":
        return value is None
    return True


class LlmCallExecutor:
    step_type = "llm.call"
    step_description = "Call an LLM with prompt/model and optional structured output."
    dsl_snippet = """
### llm.call - Call a language model
IMPORTANT: use `prompt` (NOT `messages`). `prompt` is REQUIRED. `model` is required unless the runtime injects a default model.
IMPORTANT: `temperature` and `reasoning` are optional overrides.
Omit them unless the task explicitly needs them; the runtime removes unsupported parameters based on model capabilities.
Basic call:
```yaml
- id: summarize
  type: llm.call
  input:
    model: gpt-4                        # optional when runtime defaults are configured
    prompt: "Summarize: ${data.inputs.task}"  # required - plain string
    system: "You are a helpful assistant."    # optional
    temperature: 0.7                     # optional override; omit by default
    reasoning: high                      # optional override; omit by default
    max_tokens: 2048                     # optional
```
Structured output:
IMPORTANT for `strict: true` (OpenAI/GitHub Models response_format json_schema):
- Every schema object with `properties` MUST have `required` listing EVERY key from `properties`.
- Do NOT list only the fields that feel mandatory; strict mode rejects omitted property names.
- Optional fields must still be listed in `required`; represent them as nullable with `anyOf: [{ type: <type> }, { type: "null" }]`.
- Add `additionalProperties: false` on every object schema for portability. The OpenAI provider also patches/adapts it automatically in strict mode.
```yaml
- id: classify
  type: llm.call
  input:
    model: gpt-4
    prompt: "Classify this ticket and return JSON"
    structured_output:
      schema_inline:
        type: object
        properties:
          category: { type: string }
          priority: { type: string }
          notes:
            anyOf:
              - type: string
              - type: "null"
        required: [category, priority, notes]   # every property above is listed, including nullable optional fields
        additionalProperties: false
      strict: true
```
You can also use `structured_output.schema_ref` instead of `schema_inline`.
Output: `{ text, json?, usage?, raw? }`.
"""
    documented_exceptions = [
        (ErrorCodes.INPUT_VALIDATION, False, "llm.call input/model/prompt is invalid."),
        (ErrorCodes.LLM_TIMEOUT, True, "transient timeout while calling LLM provider."),
        (ErrorCodes.LLM_NETWORK, True, "transient network failure while calling LLM provider."),
        (ErrorCodes.LLM_SCHEMA, False, "structured_output response was not valid JSON for the requested schema."),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        client = ctx.engine.llm_client
        if client is None:
            raise WorkflowRuntimeException(ErrorCodes.LLM_NETWORK, "No LLM client configured")
        input_obj = ctx.engine.get_resolved_input(ctx)
        if not isinstance(input_obj, dict):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "llm.call input must be object")

        provider, model = ctx.engine.resolve_llm_target(input_obj.get("provider"), input_obj.get("model"))
        if not model:
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "llm.call requires 'model' unless runtime default is configured")
        prompt = input_obj.get("prompt")
        if not isinstance(prompt, str):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "llm.call requires 'prompt'")

        structured = input_obj.get("structured_output") if isinstance(input_obj.get("structured_output"), dict) else None
        # Reasoning / thinking effort: "minimal"|"low"|"medium"|"high"|"max"|"auto".
        # When omitted, providers fall back to their own defaults ("auto").
        reasoning_raw = input_obj.get("reasoning")
        reasoning = reasoning_raw.strip() if isinstance(reasoning_raw, str) and reasoning_raw.strip() else None
        request = LLMRequest(
            provider=provider,
            model=model,
            prompt=prompt,
            temperature=float(input_obj["temperature"]) if "temperature" in input_obj and input_obj["temperature"] is not None else None,
            structured_output_schema=(structured.get("schema_inline") or structured.get("schema_ref")) if structured else None,
            structured_output_strict=bool(structured.get("strict")) if structured and "strict" in structured else None,
            reasoning=reasoning,
        )
        request = ctx.engine.sanitize_llm_request(request)

        ctx.set_telemetry_attribute("gen_ai.operation.name", "chat")
        ctx.set_telemetry_attribute("gen_ai.system", provider or "default")
        ctx.set_telemetry_attribute("gen_ai.request.model", model)
        if request.temperature is not None:
            ctx.set_telemetry_attribute("gen_ai.request.temperature", request.temperature)
        if request.reasoning is not None:
            ctx.set_telemetry_attribute("gen_ai.request.reasoning_effort", request.reasoning)

        if ctx.limits.log_step_content:
            ctx.add_telemetry_event(
                "gen_ai.content.prompt",
                [("gen_ai.prompt", prompt), ("prompt.role", "user")],
            )

        ctx.add_telemetry_event(
            "gnougo-flow.step.thinking",
            [
                ("gnougo-flow.thinking.message", f"Calling LLM ({model})..."),
                ("gnougo-flow.thinking.level", "thinking"),
            ],
        )

        try:
            response = await ctx.engine.call_llm_async(request)
        except TimeoutError as exc:
            raise WorkflowRuntimeException(ErrorCodes.LLM_TIMEOUT, str(exc), True) from exc
        except Exception as exc:
            raise WorkflowRuntimeException(ErrorCodes.LLM_NETWORK, f"LLM call failed: {exc}") from exc

        finish_reason = "tool_calls" if response.tool_calls else "stop"
        ctx.set_telemetry_attribute("gen_ai.response.model", model)
        ctx.set_telemetry_attribute("gen_ai.response.finish_reason", finish_reason)
        _runtime._extract_usage_telemetry(ctx, response.usage, model)

        if response.text:
            preview = response.text[:120] + ("..." if len(response.text) > 120 else "")
            ctx.add_telemetry_event(
                "gnougo-flow.step.thinking",
                [
                    ("gnougo-flow.thinking.message", preview),
                    ("gnougo-flow.thinking.level", "response"),
                ],
            )

        if ctx.limits.log_step_content and (response.text or response.json_payload is not None):
            completion_payload = response.text if response.text else json.dumps(response.json_payload, ensure_ascii=False)
            ctx.add_telemetry_event(
                "gen_ai.content.completion",
                [
                    ("gen_ai.completion", completion_payload),
                    ("completion.role", "assistant"),
                    ("completion.finish_reason", finish_reason),
                ],
            )

        structured_json = response.json_payload
        schema = request.structured_output_schema
        if schema is not None and structured_json is None and response.text:
            try:
                structured_json = json.loads(response.text)
            except Exception:
                structured_json = None
        if schema is not None and (structured_json is None or not _matches_json_schema(structured_json, schema)):
            raise WorkflowRuntimeException(
                ErrorCodes.LLM_SCHEMA,
                "llm.call structured_output expected valid JSON but the LLM returned an incompatible response",
            )

        out = {"text": response.text, "meta": {"model": model}}
        if structured_json is not None:
            out["json"] = structured_json
        if response.usage is not None:
            out["usage"] = response.usage
        if response.raw is not None:
            out["raw"] = response.raw
        return out
